File: old/ts/std/vect.ts
import * as ut from "../../util"
import * as cc from "../core"
import { union_builder_t } from "../union"
import { record_builder_t } from "../record"
import { type_t } from "../type-of-type"
import { ref_t } from "../ref"
import { nat } from "./nat"

let m = new cc.module_t ("vect")

// m.use (nat)

m.define ("vect_t", new union_builder_t ("vect_t", [
  new ref_t ("vect_null_t"),
  new ref_t ("vect_cons_t"),
], map => ({
  t: new type_t (),
  length: new ref_t ("nat_t"),
})))

m.define ("vect_null_t", new record_builder_t ("vect_null_t", map => ({
  t: new type_t (),
  length: new ref_t ("zero_t"),
})))

// m.define ("vect_cons_t", new record_builder_t ("vect_cons_t", map => ({
//   [implicit]: {
//     n: new ref_t ("nat_t"),
//   },
//   t: new type_t (),
//   length: m.game ("succ_t") .choices ({
//     prev: new this_t (map, "n"),
//   }),
//   car: new this_t (map, "t"),
//   cdr: m.game ("vect_t") .choices ({
//     t: new this_t (map, "t"),
//     length: new this_t (map, "n"),
//   }),
// })))

export { m as vect }

File: examples/vect.cs
union vect_t {
  null_vect_t
  cons_vect_t
} {
  t: type
  length: nat_t
}

class null_vect_t {
  t: type
  length: zero_t ()
}

class cons_vect_t {
  [implicit]: { n: nat_t }
  t: type
  length: succ_t (n)
  car: t
  cdr: vect_t (t, n)
}

vect_append: (
  [implicit]: {
    t: type,
    m: nat_t,
    n: nat_t,
  }
  ante: vect_t (t, m),
  succ: vect_t (t, n),
) -> vect_t (t, nat_add (m, n)) = case (ante) {
  null_vect_t => succ
  cons_vect_t => cons_vect_t (
    car = ante.car
    cdr = vect_append (ante.cdr, succ)
  )
}

vect_map: (
  [implicit]: {
    a: type,
    b: type,
    n: nat_t,
  }
  fun: (a) -> b,
  vect: vect_t (a, n),
) -> vect_t (a, n) = case (vect) {
  null_vect_t => vect
  cons_vect_t => cons_vect_t (
    car = fun (vect.car)
    cdr = vect_map (fun, vect.cdr)
  )
}
